### PR TITLE
2671372 location as entity

### PR DESCRIPTION
**Pull request by olafkarsten**
So let us begin to work on this beast.

It touches multiple issues. I will update the issues as we finish the work on this pr.

First of all it implements the stock location entity and removes the custom implementation. Thats why we have much boilerplate stuff inbetween. In multiple places we now pass location entities instead of location_ids around. That has required multiple adjustments all over the place, too.

I created an StockLocationInterface on the API Level to ensure, that we do not need the commerce stock local module for the other modules (ui to say). Probably we need to put some setters in, too. The transaction interface is not yet ready, so lets see what we need there and adjust the interface accordingly.

The PR introduces events, especially the [FILTER_STOCK_LOCATIONS](https://github.com/BBGuy/commerce_stock/compare/8.x-1.x...olafkarsten:2671372-location-as-entity?expand=1#diff-dbbc9d83eafb41a25703fc13343053fe) events. It is dispatched by the[ loadEnabled()](https://github.com/BBGuy/commerce_stock/compare/8.x-1.x...olafkarsten:2671372-location-as-entity?expand=1#diff-92eada4ee1ea7b2b588662c1bb20a13c) method of the location storage class. Test is on board for the loadEnabled() method. It returns the active locations for now. Havn't found a way to test that the event gets dispatched with the correct data. Commerce core doesn't test this. But I think with a mocked listener this should somehow work. We need to figure out Rules integration as well. We can leave both for a follow up.

Removed the getLocationsList from StockCheckInterface as we provide the locations from the StockServiceConfigInterface.

Removed the StockServiceConfig class from the API module. In my mind we don't need it. The AllwaysInStock class can return some empty arrays alone. AllwaysInStock now implements all three interfaces. One File less in the main dir. 

Problems:
~~The [StockLevelUpdater](https://github.com/BBGuy/commerce_stock/compare/8.x-1.x...olafkarsten:2671372-location-as-entity?expand=1#diff-b4d547e0b2b49c5c5cd3aa33473e4550) is broken (cron stuff). The processItem method gets entity_id for a purchasable Entity. No idea for the moment how to efficiently load the entity. But I remember there is an issue or comment about the cron stuff.~~
~~Update: Fixed it for the time beeing for products variants. There are some issues for the transaction interfaces and cron /queue handling open. Until we agree on a concept, this workes at least for product variants.~~
Update: Fixed by another commit.

~~The inline entity form throws an not so helpful error when trying to update the stock-level of the variation. Didn't find a way to debug this. No idea where it comes from.~~

### CODE REVIEW

File: modules/local_storage/commerce_stock_local.install [hunk 1]
```diff
@@ -292,4 +266,12 @@ function commerce_stock_local_install() {
       'parent_id' => 1,
     ])
     ->execute();
+
```
Comment (olafkarsten) [inline modules/local_storage/commerce_stock_local.install:276]: Creates a main stock location entity, so the users don't need to install one, if the only need one.

**Comment by olafkarsten** [comment]
Okay, I found the reason for the failing inline variation entity form. I have set the stock level field as required. That doesn't work in case of inline variation entity form, because on creation time we do not show any input element.

@steveoliver Do you know, if we have noted that allready somewhere in the issue queue?

**Comment by olafkarsten** [comment]
Green Again. Merged the new changes from the latest PRs in and solved the resulting conflicts.

File: tests/src/Functional/OrderEventTransactionsTest.php [hunk 2]
```diff
@@ -274,8 +296,10 @@ public function testWorkflowCancelEvent() {
 
   /**
    * Tests that transactions are created for all other order and order item events.
+   *
+   * @ToDo Do not pass automated tests. Investigate and fix.
    */
-  public function testOrderEvents() {
+  public function disabledTestOrderEvents() {
```
Comment (steveoliver) [inline tests/src/Functional/OrderEventTransactionsTest.php]: This has been fixed upstream.  Rebase.

**Comment by steveoliver** [comment]
Very nice! Looks just as expected. I questioned what to do about Zone, but that can definitely go in a different follow-up PR, if anything needs to be done.  I suspect we'll reconsider zone in the context of the Transaction API.  This looks great, Olaf.  @BBGuy please review.

File: tests/src/Functional/OrderEventTransactionsTest.php (hunk 2 above)
Comment (olafkarsten) [reply tests/src/Functional/OrderEventTransactionsTest.php]: done

**Comment by olafkarsten** [comment]
@BBGuy @steveoliver merged the latests commits and happily resolved the conflicts

**Comment by olafkarsten** [comment]
@bbguy Merged the new commits and resolved the conflicts, but havn't fixed the failing Tests introduced by [ProductAdminTests](https://github.com/BBGuy/commerce_stock/commit/0202d064039aaceba08fb5e8acc63effb84d7b54#diff-17e89dffc69a9804cbd5fabedb01bd96) i think.

**Comment by BBGuy** [comment]
This is a great  PR and I am exited to see location as entities. 
I did some testing and works as expected as far as I can tell. 
However I have a an issue with the removal of getLocationsList from StockCheckInterface, The removal of StockServiceConfig and the renaming of the method getLocationList() into getEnabledLocations().
Those are all open to discussion but I would expect a strong argument as to why those are essential for converting locations into entities or to go in a separate issue/PR. 
I am aware that those are topics that need attention but I really hope those can be separated to another issue and get the attention/discussion they deserve.

**Comment by olafkarsten** [comment]
Hi @BBGuy 

The code is 6 or 7 weeks old. We just started all kinds of refactoring all over the place. The module wasn't in the shape like it is now. At that point it seemed okay to solve multiple things in one PR. But you are right, from todays point of view it may have been better to split something away. Maybe we can figure the two things out, without the need to split it. 

We have a open issue for one of your concerns. Currently we have two interfaces that defines a method getLocationsList(). StockCheckInterface and StockServiceConfigInterface. And after [a short discussion](https://www.drupal.org/node/2838526) I removed the one from StockCheckInterface.

The disappeared `getLocationList()` is more a split than a simple renaming. We need 2 slightly different functions for fetching locations. `getLocations()` returns all unfiltered Locations, e.g. for admin interfaces were you want all locations not only the active/filtered ones. `getEnabledLocations()` returns the filtered locations. Per default we filter by status and return only the active ones. Since we dispatch an event, others can further filter the list. The naming is more a convention in the commerce space. You will see getENTITY(S) all over the place. I think it's better for DX if we name our methods in contrib the same. Therefor getLocations()/getEnabledLocations() and not getLocationsList()/getEnabledLocationsList()
Example: [getVariations()](https://github.com/drupalcommerce/commerce/blob/8.x-2.x/modules/product/src/Entity/ProductInterface.php#L85). 

As I write this, I'm not sure whether we should do this the other way around. Maybe `getLocations()`should return the filtered locations. The second method could be `getAllLocations()`. Why? `getVariations()` (see above) returns the filtered variations too. So maybe its better we follow that.

The removal of StockServiceConfig just happens ;). I had to refactor AllwaysInStock and StockServiceConfig to fit the new locationstuff. We would have ended up with a class (StockServiceConfig) that contains [3 methods](https://github.com/BBGuy/commerce_stock/pull/9/files#diff-433d814129dddf1ed078cdfffd0aec03R52) which return emtpy arrays. I thought that does not justify a separate class. This does not effect anything outside our own AlwaysInStock.

**Comment by BBGuy** [comment]
Hi @olafkarsten I am hesitant to give more feedback as it is a bigger discussion, however in my mind getLocationsList() currently is saying "Get locations available for stock allocation for the following purchasable entity". This is not clear and we need more work to be able to support this and that's why I am concerned about including changes and discussion in this issue.    
I am not overly concerned about AllwaysInStock. It just that as I was reviewing, I felt that it was a result of architectural changes I was not sure about. 
I think that sorting out the filtering of locations is a very important topic and one we should give attention so I created https://www.drupal.org/node/2856248
If at all possible I would love this PR to be just about location as entity.

**Comment by olafkarsten** [comment]
Hmm, you mean we should remove all the event dispatching stuff? Why? I think we discussed exactly this as one of the mean reasons to introduce the location entity at all. 1) Adding custom data to a location 2) Add an event to filter the location against whatever you want.
[I've found the issue](https://www.drupal.org/node/2838050)

Currently we return all active locations per default. The PR doesn't change that, it only allows for easier implementation of all kind of further filtering. So the issue is more a follow up, not something that we have to address right now.

Furthermore, we have to clarify the doc comments/documentation.

**Comment by BBGuy** [comment]
Yes I think I am happy with the event dispatching stuff and I like it loads. I don't think it has caused any code refactoring as this is my main concern (can check again and confirm). I would like this to be as focused as possible and only change things if needed like the obvious replacing of $location_id with $location->getId(). Adding things that don't need changing existing code is less of an issue i.e. event dispatching.

**Comment by BBGuy** [comment]
Just added a related comment https://www.drupal.org/node/2838050#comment-11961380 that explains the two types of **getLocationList()** methods the module uses.

**Comment by olafkarsten** [comment]
Created new one.
https://github.com/BBGuy/commerce_stock/pull/25

**Pull request closed without merging**